[PT-D] Prototype for megatron-lm style MLP layers

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69735
* #70079

We want to build a prototype of Megatron-LM so that we can apply PT-D op to models like transformer and other Meta flagship models like

The basic idea of Megatron-LM is as following:
1. Col-wise sharding of linear weight. Perform the linear op for the first layer.
2. Perform a math op (optional), such as ReLU or GeLU. We use GeLU in our example unit test. The input is from step 1.
3. Row-wise sharing of linear weight. Perform the linear op for the second layer. The input is from step 2.

We then save communications to concatenate the col-wise sharding results and spreading the input to different ranks for row-wise sharding.

The change is as following:
1. Return a ShardedTensor for the col-wise sharding in the sharded_linear op.
2. Return a PartialTensor for the row-wise sharding in the sharded_linear op.
3. Add helper functions to merge/aggregate local results to a fully sync local result if needed.
4. Add helper function to create sharded tensor based on the local result.
5. Add a unit test to test the Megatron-LM idea mentioned above and compare with local ops, including the grad so that we can ensure the correctness of the implementation.
6. Refactor the unit test of sharded linear to reflect the changes in the code.

Differential Revision: [D32978221](https://our.internmc.facebook.com/intern/diff/D32978221/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32978221/)!

Commit:
[PT-D] Enable megatron-lm style MLP layers (Changes mainly on sharded linear op)

We want to build a prototype of Megatron-LM so that we can apply PT-D op to models like transformer and other Meta flagship models like

The basic idea of Megatron-LM is as following:
1. Col-wise sharding of linear weight. Perform the linear op for the first layer.
2. Perform a math op (optional), such as ReLU or GeLU. We use GeLU in our example unit test. The input is from step 1.
3. Row-wise sharing of linear weight. Perform the linear op for the second layer. The input is from step 2.

We then save communications to concatenate the col-wise sharding results and spreading the input to different ranks for row-wise sharding.

The change is as following:
1. Return a ShardedTensor for the col-wise sharding in the sharded_linear op.
2. Return a list of partial tensors for the row-wise sharding in the sharded_linear op.
3. Add helper functions to merge/aggregate local results to a fully sync local result if needed.
4. Add helper function to create sharded tensor based on the local result.
5. Add a unit test to test the Megatron-LM idea mentioned above and compare with local ops, including the grad so that we can ensure the correctness of the implementation.
6. Refactor the unit test of sharded linear to reflect the changes in the code.

Differential Revision: [D32978221](https://our.internmc.facebook.com/intern/diff/D32978221/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32978221/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -510,3 +511,109 @@ def shard_parameter(
+    Given a sharded tensor and local_result from an op on top of it. We want
+    to create a new sharded tensor from the local_result so that the the next
+    op can be performed on the basis of the new sharded tensor. This can seen
+    as the last step of the first phase of the Megatron-LM style model(tensor)
+    parallelism.
```
This is a generic sharded tensor function, we shouldn't mention megatron-lm here.

fduwjj replied:
sure.

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -510,3 +511,109 @@ def shard_parameter(
     setattr(module, param_name, st)
+
+
+def init_sharded_tensor_from_local_result(
```
This function doesn't seem to belong here, it is very specific to col wise sharding. Maybe we keep it in the linear implementation?

fduwjj replied:
ok, sure.

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -510,3 +511,109 @@ def shard_parameter(
+
+    Return: new sharded tensor from the local_result.
+    """
+    sharded_weight_metadata = sharded_tensor.local_shards()[0].metadata
+    current_offsets = [0] * len(local_result.size())
+    current_offsets[result_shard_dim] = sharded_weight_metadata.shard_offsets[tensor_shard_dim]
```
This is col-wise linear specific logic.

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -510,3 +511,109 @@ def shard_parameter(
+    return new_st
+
+
+def merge_sharded_local_results(
+    sharded_local_result,
+    world_size,
+    pg=None,
+):
+    """
```
This is ok for a prototype, but for a proper API we should have something like this:

```
class ShardedTensor(...):
  def reshard(self, 
      sharding_spec: ShardingSpec, 
      output_local_tensors: bool = False) -> Union[Tensor/ShardedTensor]:
    """
    Reshards a tensor given the sharding_spec, if output_local_tensors is True. 
    It would output the local shard on  each rank instead of ShardedTensor 
     (only works for a single local shard).
    """

def reshard_output(
      module: nn.Module, 
      sharding_spec: ShardingSpec, 
      output_local_tensors: bool = False) -> nn.Module:
  """
  Replaces a module with a new module that will reshard its output.
  """

class ModuleResharder(nn.Module):
    
   def __init__(self, original_module, sharding_spec, output_local_tensors):
       self.original_module = original_module
       self.sharding_spec = sharding_spec
       self.output_local_tensors = output_local_tensors

   def forward(self, *args, **kwargs):
     output = self.original_module(*args, **kwargs)
     return output.reshard(self.sharding_spec, self.output_local_tensors)
     
# Now code for sharded linear
ChunkShardingSpec(
            dim=sharding_dim,
            placements=[
                "rank:0/cuda:0",
                "rank:1/cuda:1",
                "rank:2/cuda:2",
                "rank:3/cuda:3",
            ],
        ),

fc = nn.Linear(10, 10)
shard_parameter(fc, "weight", spec)

fc = reshard_output(fc, resharding_spec, output_local_tensors=True)

result = fc(torch.rand(10))
```

pritamdamania87 replied:
cc @wanchaol Some of the unified sharder components (ex: reshard_output) are relevant here. @fduwjj Is looking into a design of what this might look like for both ShardedTensor and potentially PartialTensor and maybe you can help with some of the brainstorming around the API too :)

fduwjj replied:
Synced offline, will send out a separate PR for the whole design and update the current PR on top of that.

fduwjj replied:
API defined in this PR:  https://github.com/pytorch/pytorch/pull/70079

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -510,3 +511,109 @@ def shard_parameter(
+    return output.t().contiguous()
+
+
+def aggregate_partial_tensor_list(
+    tensor_list,
+    rank,
+    pg=None,
+):
+    """
+    Given parital result from op on a sharded tensor. We want to aggregate the
+    partial result to create a fully synced local tensor. This can seen as the
+    last step of the second phase of the Megatron-LM style model(tensor) parallelism.
```
This is ok for a prototype, but again I think for a proper API we should have a `PartialTensor` class like `ShardedTensor` and have ops on it to reshard and aggregate appropriately.

I'd suggest writing a design doc for `PartialTensor` based on the use case we have here and then implementing it.

pritamdamania87 commented on test/distributed/_sharded_tensor/ops/test_linear.py:
```diff
@@ -87,32 +89,49 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim):
         )
+        local_bias_grad = local_linear.bias.grad
+        sharded_bias_grad = sharded_linear.bias.grad
+        # For col-wise sharding, the gradient of bias is also sharded.
+        if sharded_dim == 0:
+            dist.all_reduce(local_bias_grad)
+            local_bias_grad = local_bias_grad.narrow(
+                sharded_dim, start_pos, chunk_size
+            )
+            sharded_bias_grad = sharded_bias_grad.narrow(
+                sharded_dim, start_pos, chunk_size
+            )
```
This would be unexpected from the users point of view. The bias is not sharded, yet the grad of the bias received by the user is sharded.

I think we should narrow and apply the bias using an autograd function. The forward part of that autograd function would be exactly what we do in this PR where we narrow the bias and add it. In the backward pass we would gather the grad of the biases using this logic above. That way the user always gets a non-sharded grad for the bias.

fduwjj replied:
sure, sounds good!

pritamdamania87 commented on test/distributed/_sharded_tensor/test_megatron_prototype.py:
```diff
@@ -0,0 +1,119 @@
+        )
+        local_grad_narrowed_second = local_grad_second.narrow(1, start_pos, chunk_size)
+
+        # Test backward gradient calculation.
+        self.assertEqual(sharded_weight_first.grad, local_grad_narrowed_first)
+        self.assertEqual(sharded_weight_second.grad, local_grad_narrowed_second)
```
Add a TODO here for optimizer step.

fduwjj replied:
Will rebase on top of PR: https://github.com/pytorch/pytorch/pull/69569

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -69,22 +72,68 @@ def sharded_linear(types, args, kwargs, pg):
+        )
+
+
+def _validate_embedding_param(args, kwargs):
```
Why is this method called `_validate_embedding_param` but is used for sharded linear too?

fduwjj replied:
opps... let me change it.

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/math_ops.py:
```diff
@@ -0,0 +1,23 @@
+import torch
+
+
+def math_ops(types, args, kwargs):
```
nit: This should be called `elementwise_ops` where you can just apply the op directly on shards.

fduwjj replied:
Thanks for the suggestion for the naming.

fduwjj commented on torch/distributed/_sharding_spec/_internals.py:
```diff
@@ -150,3 +150,30 @@ def get_chunked_dim_size(dim_size, split_size, idx):
     return max(min(dim_size, split_size * (idx + 1)) - split_size * idx, 0)
+
+
+def get_chuck_sharding_params(sharding_dim_size, world_size, spec, rank):
```
opps..

pritamdamania87 commented on test/distributed/_sharded_tensor/test_megatron_prototype.py:
```diff
@@ -0,0 +1,119 @@
+
+class TestShardedTensorMegatronLinear(ShardedTensorTestBase):
+    def _run_megatron_linear(self, spec, input_size, linear_size):
+        # Use same seed.
+        torch.manual_seed(0)
+        linear_size_first = linear_size[0]
+        local_linear_first_layer = torch.nn.Linear(*linear_size_first).cuda(self.rank)
+        sharded_linear_first_layer = torch.nn.Linear(*linear_size_first)
+
+        torch.manual_seed(1024)
+        linear_size_second = linear_size[1]
+        local_linear_second_layer = torch.nn.Linear(*linear_size_second).cuda(self.rank)
+        sharded_linear_second_layer = torch.nn.Linear(*linear_size_second)
```
```
class SimpleMegatronLM():
   def __init__(self):
      self.fc1 = nn.Linear()
     self.gelu = torch.nn.GELU()
     self.fc2 = nn.Linear()

  def forward(self, inp):
    return self.fc2(self.gelu(self.fc1(inp)))
```
     
```

wanchaol replied:
second Pritam's comment above, to make it a bit reader friendly, I think we should have two models, first is `SimpleMegatronLM` model, second is a `ShardedMetagronLM`, where we shard the parameter in `init`, and run sharded computations in forward.

wanchaol commented on test/distributed/_sharded_tensor/test_megatron_prototype.py:
```diff
@@ -0,0 +1,119 @@
+        sharded_linear_second_layer = torch.nn.Linear(*linear_size_second)
+
+        # Copy the weights and bias from local linear
+        sharded_linear_first_layer.weight = torch.nn.Parameter(local_linear_first_layer.weight.detach().clone())
```
the param copy between sharded_model and a local model seems pretty popular in different tests, maybe we can make it a test util?

Commit:
Update on "[PT-D] Enable megatron-lm style MLP layers (Changes mainly on sharded linear op)"


We want to build a prototype of Megatron-LM so that we can apply PT-D op to models like transformer and other Meta flagship models like

The basic idea of Megatron-LM is as following:
1. Col-wise sharding of linear weight. Perform the linear op for the first layer.
2. Perform a math op (optional), such as ReLU or GeLU. We use GeLU in our example unit test. The input is from step 1.
3. Row-wise sharing of linear weight. Perform the linear op for the second layer. The input is from step 2.

We then save communications to concatenate the col-wise sharding results and spreading the input to different ranks for row-wise sharding.

The change is as following:
1. Return a ShardedTensor for the col-wise sharding in the sharded_linear op.
2. Return a list of partial tensors for the row-wise sharding in the sharded_linear op.
3. Add helper functions to merge/aggregate local results to a fully sync local result if needed.
4. Add helper function to create sharded tensor based on the local result.
5. Add a unit test to test the Megatron-LM idea mentioned above and compare with local ops, including the grad so that we can ensure the correctness of the implementation.
6. Refactor the unit test of sharded linear to reflect the changes in the code.

Differential Revision: [D32978221](https://our.internmc.facebook.com/intern/diff/D32978221/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32978221/)!

[ghstack-poisoned]

Commit:
Update on "[PT-D] Enable megatron-lm style MLP layers (Changes mainly on sharded linear op)"


We want to build a prototype of Megatron-LM so that we can apply PT-D op to models like transformer and other Meta flagship models like

The basic idea of Megatron-LM is as following:
1. Col-wise sharding of linear weight. Perform the linear op for the first layer.
2. Perform a math op (optional), such as ReLU or GeLU. We use GeLU in our example unit test. The input is from step 1.
3. Row-wise sharing of linear weight. Perform the linear op for the second layer. The input is from step 2.

We then save communications to concatenate the col-wise sharding results and spreading the input to different ranks for row-wise sharding.

The change is as following:
1. Return a ShardedTensor for the col-wise sharding in the sharded_linear op.
2. Return a list of partial tensors for the row-wise sharding in the sharded_linear op.
3. Add helper functions to merge/aggregate local results to a fully sync local result if needed.
4. Add helper function to create sharded tensor based on the local result.
5. Add a unit test to test the Megatron-LM idea mentioned above and compare with local ops, including the grad so that we can ensure the correctness of the implementation.
6. Refactor the unit test of sharded linear to reflect the changes in the code.

Differential Revision: [D32978221](https://our.internmc.facebook.com/intern/diff/D32978221/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32978221/)!

[ghstack-poisoned]

pritamdamania87 commented on test/distributed/_sharded_tensor/ops/test_linear.py:
```diff
@@ -90,23 +101,11 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim, merge_
-            sharded_bias_grad = sharded_bias_grad.narrow(
-                sharded_dim, start_pos, chunk_size
-            )
+        dist.all_reduce(local_bias_grad)
```
We shouldn't need this based on our discussion of how we can use an autograd function for bias calculation?

fduwjj replied:
This is still needed. As we did the same for weight. Local grad needs to be aggregated across different ranks because each rank only has one input.

pritamdamania87 commented on test/distributed/_sharded_tensor/test_megatron_prototype.py:
```diff
@@ -35,72 +44,165 @@
+        def forward(self, inp):
+            return self.fc2(self.gelu(self.fc1(inp)))
+
+        def weight_override(self, module):
+            self.fc1.weight = clone_module_parameter(module.fc1, "weight")
+            self.fc1.bias = clone_module_parameter(module.fc1, "bias")
+            self.fc2.weight = clone_module_parameter(module.fc2, "weight")
+            self.fc2.bias = clone_module_parameter(module.fc2, "bias")
+
+        def shard_parameter(self, spec):
+            shard_parameter(self.fc1, "weight", spec[0])
+            shard_parameter(self.fc2, "weight", spec[1])
+
+        def get_weight_grad(self):
+            return (self.fc1.weight.grad, self.fc2.weight.grad)
+
+        def get_bias_grad(self):
+            return (self.fc1.bias.grad, self.fc2.bias.grad)
+
+        def get_weights(self):
+            return (self.fc1.weight, self.fc2.weight)
+
+        def get_bias(self):
+            return (self.fc1.bias, self.fc2.bias)
+
+        def get_weight_local_shard(self):
+            return (
+                self.fc1.weight.local_shards()[0].tensor,
+                self.fc2.weight.local_shards()[0].tensor,
+            )
```
All of this should not be part of the model. Let us put these helper methods elsewhere. In general the philosophy is that the model should only contain modeling code and nothing else. For example, someone writing a Transformer model should not worry about sharding, weight override etc. Our sharding APIs should provide all this.

fduwjj replied:
Ahh, I see. Will do.

pritamdamania87 commented on test/distributed/_sharded_tensor/test_megatron_prototype.py:
```diff
@@ -35,72 +44,165 @@
+        self.assertEqual(bias_fc1, local_bias_fc1)
+        self.assertNotEqual(previous_bias_fc2, bias_fc2)
+        # Sharded bias get double updated here for fc2.
+        # self.assertEqual(bias_fc2, local_bias_fc2)
```
cc @wanchaol

fduwjj replied:
Yes, this is very wired bug which I don't know the reason. Because all grads and updated weights are working as expected. Even the bias update for fc1 is correct.

wanchaol replied:
Yeah debugged this and found the issue, it should be fixed by https://github.com/pytorch/pytorch/pull/70228

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/elementwise_ops.py:
```diff
@@ -0,0 +1,24 @@
+    if isinstance(input, torch.Tensor):
+        return op(args[0])
+    else:
+        for local_shard in input.local_shards():
+            local_shard.tensor = op(local_shard.tensor)
```
Doesn't seem like this is the right semantics. We are modifying the ShardedTensor in place instead of returning a new ShardedTensor. For example `gelu` on a regular `torch.Tensor` doesn't modify the Tensor in-place but returns a new Tensor.

fduwjj replied:
I see.. Let me make sure the behavior is the same.

wanchaol replied:
yeah think we should return a new ShardedTensor here

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -257,17 +262,93 @@ def _handle_row_wise_sharding_tensor(
+    return new_st
+
+
+class _BiasTensorNarrow(Function):
```
nit: Add some docs for this explaining why we need it

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -257,17 +262,93 @@ def _handle_row_wise_sharding_tensor(
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        return (None, grad_output)
```
Do we really need this? The backward part of this function is not doing anything useful and just forwarding the grad?

fduwjj replied:
This is really tricky, I found that if I don't use this one, the grad will be incorrect. I will try to debug more here.

fduwjj replied:
I think I find why this is needed. Like `_BiasTensorNarrow`, in the forward, we multiply 1/4 (1/world_size) but during backward, we don't need that. That's why we need to manually skip the div during backward.

pritamdamania87 reviewed (commented):
Awesome stuff! Great to see everything actually works with the new proposed APIs!

wanchaol commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -74,20 +85,66 @@ def sharded_linear(types, args, kwargs, pg):
+
+def _validate_linear_param(args, kwargs):
+    """
+    Validate input params of sharded embedding op.
```
nit: linear op

fduwjj commented on test/distributed/_sharded_tensor/ops/test_linear.py:
```diff
@@ -64,12 +75,14 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim):
         sharded_output = torch.nn.functional.linear(
             inp, sharded_linear.weight, sharded_linear.bias
         )
+        sharded_output = sharded_output.reshard(resharding_spec)
```
Because we call linear API here. And it has not been wrapped.

Commit:
Update on "[PT-D] Enable megatron-lm style MLP layers (Changes mainly on sharded linear op)"


We want to build a prototype of Megatron-LM so that we can apply PT-D op to models like transformer and other Meta flagship models like

The basic idea of Megatron-LM is as following:
1. Col-wise sharding of linear weight. Perform the linear op for the first layer.
2. Perform a math op (optional), such as ReLU or GeLU. We use GeLU in our example unit test. The input is from step 1.
3. Row-wise sharing of linear weight. Perform the linear op for the second layer. The input is from step 2.

We then save communications to concatenate the col-wise sharding results and spreading the input to different ranks for row-wise sharding.

The change is as following:
1. Return a ShardedTensor for the col-wise sharding in the sharded_linear op.
2. Return a list of partial tensors for the row-wise sharding in the sharded_linear op.
3. Add helper functions to merge/aggregate local results to a fully sync local result if needed.
4. Add helper function to create sharded tensor based on the local result.
5. Add a unit test to test the Megatron-LM idea mentioned above and compare with local ops, including the grad so that we can ensure the correctness of the implementation.
6. Refactor the unit test of sharded linear to reflect the changes in the code.

Differential Revision: [D32978221](https://our.internmc.facebook.com/intern/diff/D32978221/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32978221/)!

[ghstack-poisoned]

fduwjj commented:
1. Address comments from reviewers.
2. Add more comments and unit test.

Commit:
Update on "[PT-D] Enable megatron-lm style MLP layers (Changes mainly on sharded linear op)"


We want to build a prototype of Megatron-LM so that we can apply PT-D op to models like transformer and other Meta flagship models like

The basic idea of Megatron-LM is as following:
1. Col-wise sharding of linear weight. Perform the linear op for the first layer.
2. Perform a math op (optional), such as ReLU or GeLU. We use GeLU in our example unit test. The input is from step 1.
3. Row-wise sharing of linear weight. Perform the linear op for the second layer. The input is from step 2.

We then save communications to concatenate the col-wise sharding results and spreading the input to different ranks for row-wise sharding.

The change is as following:
1. Return a ShardedTensor for the col-wise sharding in the sharded_linear op.
2. Return a list of partial tensors for the row-wise sharding in the sharded_linear op.
3. Add helper functions to merge/aggregate local results to a fully sync local result if needed.
4. Add helper function to create sharded tensor based on the local result.
5. Add a unit test to test the Megatron-LM idea mentioned above and compare with local ops, including the grad so that we can ensure the correctness of the implementation.
6. Refactor the unit test of sharded linear to reflect the changes in the code.

Differential Revision: [D32978221](https://our.internmc.facebook.com/intern/diff/D32978221/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32978221/)!

[ghstack-poisoned]

fduwjj commented:
Address mypy feedback and rebase.

Commit:
Update on "[PT-D] Enable megatron-lm style MLP layers (Changes mainly on sharded linear op)"


We want to build a prototype of Megatron-LM so that we can apply PT-D op to models like transformer and other Meta flagship models like

The basic idea of Megatron-LM is as following:
1. Col-wise sharding of linear weight. Perform the linear op for the first layer.
2. Perform a math op (optional), such as ReLU or GeLU. We use GeLU in our example unit test. The input is from step 1.
3. Row-wise sharing of linear weight. Perform the linear op for the second layer. The input is from step 2.

We then save communications to concatenate the col-wise sharding results and spreading the input to different ranks for row-wise sharding.

The change is as following:
1. Return a ShardedTensor for the col-wise sharding in the sharded_linear op.
2. Return a list of partial tensors for the row-wise sharding in the sharded_linear op.
3. Add helper functions to merge/aggregate local results to a fully sync local result if needed.
4. Add helper function to create sharded tensor based on the local result.
5. Add a unit test to test the Megatron-LM idea mentioned above and compare with local ops, including the grad so that we can ensure the correctness of the implementation.
6. Refactor the unit test of sharded linear to reflect the changes in the code.

Differential Revision: [D32978221](https://our.internmc.facebook.com/intern/diff/D32978221/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32978221/)!

[ghstack-poisoned]

fduwjj commented:
Put the unit test into the windows block list.

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/elementwise_ops.py:
```diff
@@ -0,0 +1,35 @@
+    # Validate types
+    if not isinstance(input, torch.Tensor) and not isinstance(input, ShardedTensor):
+        raise TypeError("input needs to be either torch.Tensor or ShardedTensor")
+    if isinstance(input, torch.Tensor):
+        return op(args[0])
```
Do we need to handle this case at all? Shouldn't normal tensor dispatch take care of this? My understanding is that this codepath will never be executed given how `__torch_function__` dispatch works.

fduwjj replied:
I see.. That makes sense. Thanks for pointing that out.

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/elementwise_ops.py:
```diff
@@ -0,0 +1,35 @@
+        raise TypeError("input needs to be either torch.Tensor or ShardedTensor")
+    if isinstance(input, torch.Tensor):
+        return op(args[0])
+    else:
```
nit: We don't really need an `else:` here since above we either raise or just return.

fduwjj replied:
Sure.

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/elementwise_ops.py:
```diff
@@ -0,0 +1,35 @@
+        local_shards_new = []
+        for local_shard in input.local_shards():
+            local_shards_new.append(Shard(op(local_shard.tensor), local_shard.metadata))
+        new_st = ShardedTensor._init_from_local_shards(
+            local_shards_new, input.size(), process_group=pg
+        )
```
This method actually performs an allgather: https://github.com/pytorch/pytorch/blob/master/torch/distributed/_sharded_tensor/api.py#L351. We probably need a variation of this that doesn't do the validation for perf reasons or have a separate simpler API.

For now, this is fine but we should improve upon this in follow up PRs once we have an API to do this efficiently.

cc @wanchaol

fduwjj replied:
As we discussed offline, will create a new API named `_init_from_local_tensor` to avoid the potential cost of `all_gather_object`.

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/elementwise_ops.py:
```diff
@@ -0,0 +1,35 @@
+    if isinstance(input, torch.Tensor):
+        return op(args[0])
+    else:
+        local_shards_new = []
+        for local_shard in input.local_shards():
+            local_shards_new.append(Shard(op(local_shard.tensor), local_shard.metadata))
+        new_st = ShardedTensor._init_from_local_shards(
+            local_shards_new, input.size(), process_group=pg
+        )
```
This logic is probably common across all elementwise ops. To make this generic, can we support both `gelu` and `relu` and have them both share the same code? In an ideal world, just adding one line of code should be sufficient to add a new elementwise op:

```
register_elementwise_op(torch.nn.functional.gelu)
register_elementwise_op(torch.nn.functional.relu)
```

fduwjj replied:
Sure. Let me make this a decorator so that we can easily support `elementwise_op` in a just one line.

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -138,28 +175,40 @@ def _handle_col_wise_sharding(input, world_size, weight, local_shard_t, bias, pg
+    local_bias = _BiasTensorNarrow.apply(
+        world_size, start_pos, chunk_size, weight, pg, bias
+    )
+    results = [torch.tensor(1)] * world_size
```
Why can't this just be `None`?

fduwjj replied:
Will make it None and add a mypy suppression.

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -218,15 +268,144 @@ def _handle_row_wise_sharding(input, world_size, weight, rank, local_shard_t, bi
+        placement=sharded_weight_metadata.placement,
+    )
+    local_shards = [Shard(local_result, local_shard_metadata)]
+    new_st = ShardedTensor._init_from_local_shards(
```
This has additional allgather overhead as mentioned earlier: https://github.com/pytorch/pytorch/blob/master/torch/distributed/_sharded_tensor/api.py#L351

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/embedding.py:
```diff
@@ -104,7 +104,7 @@ def sharded_embedding(types, args, kwargs, pg):
     norm_type = kwargs.get("norm_type")
```
After this PR, the linear and elementwise ops produce sharded results, but the embedding ops still collect all the results.

For consistency, we should have a follow up PR such that embedding ops also produce sharded results.

fduwjj replied:
Put this requirement in the gh issue: https://github.com/pytorch/pytorch/issues/72022

pritamdamania87 commented on test/distributed/_sharded_tensor/ops/test_linear.py:
```diff
@@ -44,24 +48,32 @@
+            _reshard_output(sharded_linear, reshard_spec)
+        )
         sharded_output = sharded_linear(inp)
+        if unstack_output:
+            sharded_output = torch.cat(torch.unbind(sharded_output))
```
Why do we need to do this? Try to write unit tests from the users point of view to see if the APIs we introduced are user friendly or not.

For example, the idea of adding `_collect_local_shard` and `_reshard_output` was to ensure the user doesn't have to do additional operations like this to get the desired result.

fduwjj replied:
This is because we have this tricky corner case when the input is only one dimension. The col-wise sharding will concatenate all local results also in the dimension 0. I tried a workaround to make it `stack` instead, that's why we need an `unbind` here. Let me think more on this to see if we can have a cleaner solution.

fduwjj replied:
Another work-around is that for this type of case, we will make the shard tensor shard by dim -1. So that we can set the dim back to the last dim during the reshard work. (Because by default when sharding_dim = = reshard_dim, for this case both == 0, we just reshuffle the local shard). What do you think?

pritamdamania87 commented on test/distributed/_sharded_tensor/test_megatron_prototype.py:
```diff
@@ -0,0 +1,222 @@
+
+        def _get_weight_local_shard(module):
+            return (
+                module.fc1.weight.local_shards()[0].tensor,
```
nit: You can use the new `local_shard()` api :)

fduwjj replied:
Sure.

pritamdamania87 commented on test/distributed/_sharded_tensor/test_megatron_prototype.py:
```diff
@@ -0,0 +1,222 @@
+    sys.exit(0)
+
+
+class TestShardedTensorMegatronLinear(ShardedTensorTestBase):
```
Awesome unit test!

Commit:
Update on "[PT-D] Enable megatron-lm style MLP layers (Changes mainly on sharded linear op)"


We want to build a prototype of Megatron-LM so that we can apply PT-D op to models like transformer and other Meta flagship models like

The basic idea of Megatron-LM is as following:
1. Col-wise sharding of linear weight. Perform the linear op for the first layer.
2. Perform a math op (optional), such as ReLU or GeLU. We use GeLU in our example unit test. The input is from step 1.
3. Row-wise sharing of linear weight. Perform the linear op for the second layer. The input is from step 2.

We then save communications to concatenate the col-wise sharding results and spreading the input to different ranks for row-wise sharding.

The change is as following:
1. Return a ShardedTensor for the col-wise sharding in the sharded_linear op.
2. Return a list of partial tensors for the row-wise sharding in the sharded_linear op.
3. Add helper functions to merge/aggregate local results to a fully sync local result if needed.
4. Add helper function to create sharded tensor based on the local result.
5. Add a unit test to test the Megatron-LM idea mentioned above and compare with local ops, including the grad so that we can ensure the correctness of the implementation.
6. Refactor the unit test of sharded linear to reflect the changes in the code.

Differential Revision: [D32978221](https://our.internmc.facebook.com/intern/diff/D32978221/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32978221/)!

[ghstack-poisoned]

Commit:
Update on "[PT-D] Enable megatron-lm style MLP layers (Changes mainly on sharded linear op)"


We want to build a prototype of Megatron-LM so that we can apply PT-D op to models like transformer and other Meta flagship models like

The basic idea of Megatron-LM is as following:
1. Col-wise sharding of linear weight. Perform the linear op for the first layer.
2. Perform a math op (optional), such as ReLU or GeLU. We use GeLU in our example unit test. The input is from step 1.
3. Row-wise sharing of linear weight. Perform the linear op for the second layer. The input is from step 2.

We then save communications to concatenate the col-wise sharding results and spreading the input to different ranks for row-wise sharding.

The change is as following:
1. Return a ShardedTensor for the col-wise sharding in the sharded_linear op.
2. Return a list of partial tensors for the row-wise sharding in the sharded_linear op.
3. Add helper functions to merge/aggregate local results to a fully sync local result if needed.
4. Add helper function to create sharded tensor based on the local result.
5. Add a unit test to test the Megatron-LM idea mentioned above and compare with local ops, including the grad so that we can ensure the correctness of the implementation.
6. Refactor the unit test of sharded linear to reflect the changes in the code.

Differential Revision: [D32978221](https://our.internmc.facebook.com/intern/diff/D32978221/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32978221/)!

[ghstack-poisoned]

fduwjj commented:
Addressed the comment from reviewers.

Commit:
Update on "[PT-D] Prototype for megatron-lm style MLP layers "


We want to build a prototype of Megatron-LM so that we can apply PT-D op to models like transformer and other Meta flagship models like

The basic idea of Megatron-LM is as following:
1. Col-wise sharding of linear weight. Perform the linear op for the first layer.
2. Perform a math op (optional), such as ReLU or GeLU. We use GeLU in our example unit test. The input is from step 1.
3. Row-wise sharing of linear weight. Perform the linear op for the second layer. The input is from step 2.

We then save communications to concatenate the col-wise sharding results and spreading the input to different ranks for row-wise sharding.

The change is as following:
1. Return a ShardedTensor for the col-wise sharding in the sharded_linear op.
2. Return a PartialTensor for the row-wise sharding in the sharded_linear op.
3. Add helper functions to merge/aggregate local results to a fully sync local result if needed.
4. Add helper function to create sharded tensor based on the local result.
5. Add a unit test to test the Megatron-LM idea mentioned above and compare with local ops, including the grad so that we can ensure the correctness of the implementation.
6. Refactor the unit test of sharded linear to reflect the changes in the code.

Differential Revision: [D32978221](https://our.internmc.facebook.com/intern/diff/D32978221/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32978221/)!

[ghstack-poisoned]

fduwjj commented:
Synced offline with @pritamdamania87 and made the handling of corner case in `_collect_local_shard` and add comments to clarify the necessity of this special logic.

Commit:
Update on "[PT-D] Prototype for megatron-lm style MLP layers "


We want to build a prototype of Megatron-LM so that we can apply PT-D op to models like transformer and other Meta flagship models like

The basic idea of Megatron-LM is as following:
1. Col-wise sharding of linear weight. Perform the linear op for the first layer.
2. Perform a math op (optional), such as ReLU or GeLU. We use GeLU in our example unit test. The input is from step 1.
3. Row-wise sharing of linear weight. Perform the linear op for the second layer. The input is from step 2.

We then save communications to concatenate the col-wise sharding results and spreading the input to different ranks for row-wise sharding.

The change is as following:
1. Return a ShardedTensor for the col-wise sharding in the sharded_linear op.
2. Return a PartialTensor for the row-wise sharding in the sharded_linear op.
3. Add helper functions to merge/aggregate local results to a fully sync local result if needed.
4. Add helper function to create sharded tensor based on the local result.
5. Add a unit test to test the Megatron-LM idea mentioned above and compare with local ops, including the grad so that we can ensure the correctness of the implementation.
6. Refactor the unit test of sharded linear to reflect the changes in the code.

Differential Revision: [D32978221](https://our.internmc.facebook.com/intern/diff/D32978221/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32978221/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -602,6 +602,19 @@ def _collect_local_shard(module: torch.nn.Module) -> torch.nn.Module:
     """
     Hook a module with local shards collection in the forward pass.
 
+    When the local tensor only has one dimension, we increase one more dimension
+    for reshard. We then do squeeze here manually to reduce the dimension.
```
I don't think we should specify reshard or sharded linear in the docs of this method. I think the docs should be something like this:

```
This API is typically used to convert a sharded representation back to data parallel representation. In particular, it returns the local tensor for this Shard. If the size along the sharding dimension for the local tensor is 1, this dimension is removed from the final result. For example a [4, 16] ShardedTensor across 4 ranks is typically a local Tensor of size [16] across each rank and not [1, 16] across each rank.
```

fduwjj replied:
sure.

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/elementwise_ops.py:
```diff
@@ -33,3 +30,7 @@ def gelu(types, args=(), kwargs=None, pg=None):
         return new_st
+
+
+gelu = register_elementwise_op(torch.nn.functional.gelu)
+relu = register_elementwise_op(torch.nn.functional.relu)
```
nit: Don't need `gelu = register_elementwise_op(torch.nn.functional.gelu)`, you can simply do `register_elementwise_op(torch.nn.functional.gelu)`?

fduwjj replied:
That works, thanks!

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/elementwise_ops.py:
```diff
@@ -8,21 +8,18 @@
+            raise TypeError("input needs to be a ShardedTensor")
         local_shards_new = []
         for local_shard in input.local_shards():
             local_shards_new.append(Shard(op(local_shard.tensor), local_shard.metadata))
```
Lets add a TODO here to use the `from_local_tensor` API to avoid allgather overhead.

fduwjj replied:
Put in this issue: https://github.com/pytorch/pytorch/issues/72092

Commit:
Update on "[PT-D] Prototype for megatron-lm style MLP layers "


We want to build a prototype of Megatron-LM so that we can apply PT-D op to models like transformer and other Meta flagship models like

The basic idea of Megatron-LM is as following:
1. Col-wise sharding of linear weight. Perform the linear op for the first layer.
2. Perform a math op (optional), such as ReLU or GeLU. We use GeLU in our example unit test. The input is from step 1.
3. Row-wise sharing of linear weight. Perform the linear op for the second layer. The input is from step 2.

We then save communications to concatenate the col-wise sharding results and spreading the input to different ranks for row-wise sharding.

The change is as following:
1. Return a ShardedTensor for the col-wise sharding in the sharded_linear op.
2. Return a PartialTensor for the row-wise sharding in the sharded_linear op.
3. Add helper functions to merge/aggregate local results to a fully sync local result if needed.
4. Add helper function to create sharded tensor based on the local result.
5. Add a unit test to test the Megatron-LM idea mentioned above and compare with local ops, including the grad so that we can ensure the correctness of the implementation.
6. Refactor the unit test of sharded linear to reflect the changes in the code.

Differential Revision: [D32978221](https://our.internmc.facebook.com/intern/diff/D32978221/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32978221/)!

[ghstack-poisoned]

fduwjj commented:
Rebased on top of https://github.com/pytorch/pytorch/pull/72141#pullrequestreview-870034365.

Commit:
Update on "[PT-D] Prototype for megatron-lm style MLP layers "


We want to build a prototype of Megatron-LM so that we can apply PT-D op to models like transformer and other Meta flagship models like

The basic idea of Megatron-LM is as following:
1. Col-wise sharding of linear weight. Perform the linear op for the first layer.
2. Perform a math op (optional), such as ReLU or GeLU. We use GeLU in our example unit test. The input is from step 1.
3. Row-wise sharing of linear weight. Perform the linear op for the second layer. The input is from step 2.

We then save communications to concatenate the col-wise sharding results and spreading the input to different ranks for row-wise sharding.

The change is as following:
1. Return a ShardedTensor for the col-wise sharding in the sharded_linear op.
2. Return a PartialTensor for the row-wise sharding in the sharded_linear op.
3. Add helper functions to merge/aggregate local results to a fully sync local result if needed.
4. Add helper function to create sharded tensor based on the local result.
5. Add a unit test to test the Megatron-LM idea mentioned above and compare with local ops, including the grad so that we can ensure the correctness of the implementation.
6. Refactor the unit test of sharded linear to reflect the changes in the code.

Differential Revision: [D32978221](https://our.internmc.facebook.com/intern/diff/D32978221/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32978221/)!

[ghstack-poisoned]
